components/places/src/bookmark_sync/store.rs

```diff
@@ -515,6 +515,76 @@ impl<'a> BookmarksStore<'a> {
         Ok(())
     }
 
+    fn store_incoming_livemark(&self, modified: ServerTimestamp, l: LivemarkRecord) -> Result<()> {
+        // livemarks don't store a reference to the place, so we validate it manually.
+        fn validate_href(h: Option<String>, guid: &SyncGuid, what: &str) -> Option<String> {
+            match h {
+                Some(h) => match Url::parse(&h) {
+                    Ok(url) => {
+                        let s = url.to_string();
+                        if s.len() > URL_LENGTH_MAX {
+                            log::warn!("Livemark {} has a {} URL which is too long", &guid, what);
+                            None
+                        } else {
+                            Some(s)
+                        }
+                    },
+                    Err(e) => {
+                        log::warn!("Livemark {} has an invalid {} URL '{}'", &guid, what, h);
+                        None
+                    }
+                },
+                None => {
+                    log::warn!("Livemark {} has no {} URL", &guid, what);
+                    None
+                }
+            }
+        }
+        let feed_url = validate_href(l.feed_url, &l.guid, "feed");
+        let site_url = validate_href(l.site_url, &l.guid, "site");
+        let validity = if feed_url.is_some() {
+            SyncedBookmarkValidity::Valid
+        } else {
+            SyncedBookmarkValidity::Replace
+
+        };
+        self.db.execute_named_cached(
+            r#"REPLACE INTO moz_bookmarks_synced(guid, parentGuid, serverModified, needsMerge, kind,
+                                                 dateAdded, title, feedURL, siteURL, validity)
+               VALUES(:guid, :parentGuid, :serverModified, 1, :kind,
+                      :dateAdded, :title, :feedUrl, :siteUrl, :validity)"#,
+            &[
+                (":guid", &l.guid.as_ref()),
+                (":parentGuid", &l.parent_guid.as_ref()),
+                (":serverModified", &(modified.as_millis() as i64)),
+                (":kind", &SyncedBookmarkKind::Livemark),
+                (":dateAdded", &l.date_added),
+                (":title", &l.title),
+                (":feedUrl", &feed_url),
+                (":siteUrl", &site_url),
+                (":validity", &validity),
+            ],
+        )?;
+        Ok(())
+    }
+
+    fn store_incoming_sep(&self, modified: ServerTimestamp, s: SeparatorRecord) -> Result<()> {
+        self.db.execute_named_cached(
+            r#"REPLACE INTO moz_bookmarks_synced(guid, parentGuid, serverModified, needsMerge, kind,
```

Comment:
I was wondering if we should store the `position` from the record, too. On the one hand, newer Desktops (Dogear and the JS buffered engine) won't use it. On the other, it _might_ come in handy for fixing up orphaned separators—assuming their `parentid` exists—and we could still flag separators with the wrong `pos` for reupload.

Speaking of orphans, though...I wonder if it's better to just delete orphaned separators entirely, instead of trying to do fancy things with their `parentid` and `pos`. Moving separators to the end of the folder, or to `unfiled`, is unlikely to be useful, even if it is technically data loss.